// coc-helper/tests/coc.nvim/lib/markdown/styles.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.blue = exports.green = exports.yellow = exports.italic = exports.strikethrough = exports.underline = exports.bold = exports.magenta = exports.gray = void 0;
const node_1 = require("../util/node");
function gray(str) {
    return `${node_1.styles.gray.open}${str}${node_1.styles.gray.close}`;
}
exports.gray = gray;
function magenta(str) {
    return `${node_1.styles.magenta.open}${str}${node_1.styles.magenta.close}`;
}
exports.magenta = magenta;
function bold(str) {
    return `${node_1.styles.bold.open}${str}${node_1.styles.bold.close}`;
}
exports.bold = bold;
function underline(str) {
    return `${node_1.styles.underline.open}${str}${node_1.styles.underline.close}`;
}
exports.underline = underline;
function strikethrough(str) {
    return `${node_1.styles.strikethrough.open}${str}${node_1.styles.strikethrough.close}`;
}
exports.strikethrough = strikethrough;
function italic(str) {
    return `${node_1.styles.italic.open}${str}${node_1.styles.italic.close}`;
}
exports.italic = italic;
function yellow(str) {
    return `${node_1.styles.yellow.open}${str}${node_1.styles.yellow.close}`;
}
exports.yellow = yellow;
function green(str) {
    return `${node_1.styles.green.open}${str}${node_1.styles.green.close}`;
}
exports.green = green;
function blue(str) {
    return `${node_1.styles.blue.open}${str}${node_1.styles.blue.close}`;
}
exports.blue = blue;
