// coc-helper/tests/coc.nvim/lib/list/configuration.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListConfiguration = exports.validKeys = void 0;
const tslib_1 = require("tslib");
const window_1 = tslib_1.__importDefault(require("../window"));
const workspace_1 = tslib_1.__importDefault(require("../workspace"));
exports.validKeys = [
    '<esc>',
    '<space>',
    '<tab>',
    '<s-tab>',
    '<bs>',
    '<right>',
    '<left>',
    '<up>',
    '<down>',
    '<home>',
    '<end>',
    '<cr>',
    '<FocusGained>',
    '<FocusLost>',
    '<ScrollWheelUp>',
    '<ScrollWheelDown>',
    '<LeftMouse>',
    '<LeftDrag>',
    '<LeftRelease>',
    '<2-LeftMouse>',
    '<C-a>',
    '<C-b>',
    '<C-c>',
    '<C-d>',
    '<C-e>',
    '<C-f>',
    '<C-g>',
    '<C-h>',
    '<C-i>',
    '<C-j>',
    '<C-k>',
    '<C-l>',
    '<C-m>',
    '<C-n>',
    '<C-o>',
    '<C-p>',
    '<C-q>',
    '<C-r>',
    '<C-s>',
    '<C-t>',
    '<C-u>',
    '<C-v>',
    '<C-w>',
    '<C-x>',
    '<C-y>',
    '<C-z>',
    '<A-a>',
    '<A-b>',
    '<A-c>',
    '<A-d>',
    '<A-e>',
    '<A-f>',
    '<A-g>',
    '<A-h>',
    '<A-i>',
    '<A-j>',
    '<A-k>',
    '<A-l>',
    '<A-m>',
    '<A-n>',
    '<A-o>',
    '<A-p>',
    '<A-q>',
    '<A-r>',
    '<A-s>',
    '<A-t>',
    '<A-u>',
    '<A-v>',
    '<A-w>',
    '<A-x>',
    '<A-y>',
    '<A-z>',
];
class ListConfiguration {
    get debounceTime() {
        return this.get('interactiveDebounceTime', 100);
    }
    get extendedSearchMode() {
        return this.get('extendedSearchMode', true);
    }
    get smartcase() {
        return this.get('smartCase', false);
    }
    get signOffset() {
        return this.get('signOffset', 900);
    }
    get(key, defaultValue) {
        let configuration = workspace_1.default.initialConfiguration;
        return configuration.get('list.' + key, defaultValue);
    }
    get previousKey() {
        return this.fixKey(this.get('previousKeymap', '<C-j>'));
    }
    get nextKey() {
        return this.fixKey(this.get('nextKeymap', '<C-k>'));
    }
    fixKey(key) {
        if (exports.validKeys.includes(key))
            return key;
        let find = exports.validKeys.find(s => s.toLowerCase() == key.toLowerCase());
        if (find)
            return find;
        void window_1.default.showErrorMessage(`Configured key "${key}" not supported.`);
        return null;
    }
}
exports.ListConfiguration = ListConfiguration;
exports.default = new ListConfiguration();
