// coc-helper/tests/coc.nvim/lib/list/source/services.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const services_1 = tslib_1.__importDefault(require("../../services"));
const util_1 = require("../../util");
const basic_1 = tslib_1.__importDefault(require("../basic"));
const formatting_1 = require("../formatting");
class ServicesList extends basic_1.default {
    constructor() {
        super();
        this.defaultAction = 'toggle';
        this.description = 'registered services of coc.nvim';
        this.name = 'services';
        this.addAction('toggle', async (item) => {
            let { id } = item.data;
            await services_1.default.toggle(id);
            await (0, util_1.wait)(50);
        }, { persist: true, reload: true });
    }
    async loadItems(_context) {
        let stats = services_1.default.getServiceStats();
        return (0, formatting_1.formatListItems)(this.alignColumns, stats.map(stat => {
            let prefix = stat.state == 'running' ? '*' : ' ';
            return {
                label: [prefix, stat.id, `[${stat.state}]`, stat.languageIds.join(', ')],
                data: { id: stat.id }
            };
        }));
    }
    doHighlight() {
        let { nvim } = this;
        nvim.pauseNotification();
        nvim.command('syntax match CocServicesPrefix /\\v^./ contained containedin=CocServicesLine', true);
        nvim.command('syntax match CocServicesName /\\v%3c\\S+/ contained containedin=CocServicesLine', true);
        nvim.command('syntax match CocServicesStat /\\v\\t\\[\\w+\\]/ contained containedin=CocServicesLine', true);
        nvim.command('syntax match CocServicesLanguages /\\v(\\])@<=.*$/ contained containedin=CocServicesLine', true);
        nvim.command('highlight default link CocServicesPrefix Special', true);
        nvim.command('highlight default link CocServicesName Type', true);
        nvim.command('highlight default link CocServicesStat Statement', true);
        nvim.command('highlight default link CocServicesLanguages Comment', true);
        nvim.resumeNotification(false, true);
    }
}
exports.default = ServicesList;
