// coc-helper/tests/coc.nvim/lib/language-client/typeHierarchy.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeHierarchyFeature = void 0;
const tslib_1 = require("tslib");
const languages_1 = tslib_1.__importDefault(require("../languages"));
const protocol_1 = require("../util/protocol");
const features_1 = require("./features");
const cv = tslib_1.__importStar(require("./utils/converter"));
class TypeHierarchyFeature extends features_1.TextDocumentLanguageFeature {
    constructor(client) {
        super(client, protocol_1.TypeHierarchyPrepareRequest.type);
    }
    fillClientCapabilities(capabilities) {
        const capability = (0, features_1.ensure)((0, features_1.ensure)(capabilities, 'textDocument'), 'typeHierarchy');
        capability.dynamicRegistration = true;
    }
    initialize(capabilities, documentSelector) {
        const [id, options] = this.getRegistration(documentSelector, capabilities.typeHierarchyProvider);
        if (!id || !options) {
            return;
        }
        this.register({ id, registerOptions: options });
    }
    registerLanguageProvider(options) {
        const client = this._client;
        const selector = options.documentSelector;
        const provider = {
            prepareTypeHierarchy: (document, position, token) => {
                const prepareTypeHierarchy = (document, position, token) => {
                    const params = cv.asTextDocumentPositionParams(document, position);
                    return this.sendRequest(protocol_1.TypeHierarchyPrepareRequest.type, params, token);
                };
                const middleware = client.middleware;
                return middleware.prepareTypeHierarchy
                    ? middleware.prepareTypeHierarchy(document, position, token, prepareTypeHierarchy)
                    : prepareTypeHierarchy(document, position, token);
            },
            provideTypeHierarchySupertypes: (item, token) => {
                const provideTypeHierarchySupertypes = (item, token) => {
                    return this.sendRequest(protocol_1.TypeHierarchySupertypesRequest.type, { item }, token);
                };
                const middleware = client.middleware;
                return middleware.provideTypeHierarchySupertypes
                    ? middleware.provideTypeHierarchySupertypes(item, token, provideTypeHierarchySupertypes)
                    : provideTypeHierarchySupertypes(item, token);
            },
            provideTypeHierarchySubtypes: (item, token) => {
                const provideTypeHierarchySubtypes = (item, token) => {
                    return this.sendRequest(protocol_1.TypeHierarchySubtypesRequest.type, { item }, token);
                };
                const middleware = client.middleware;
                return middleware.provideTypeHierarchySubtypes
                    ? middleware.provideTypeHierarchySubtypes(item, token, provideTypeHierarchySubtypes)
                    : provideTypeHierarchySubtypes(item, token);
            }
        };
        return [languages_1.default.registerTypeHierarchyProvider(selector, provider), provider];
    }
}
exports.TypeHierarchyFeature = TypeHierarchyFeature;
