// coc-helper/tests/coc.nvim/lib/provider/callHierarchyManager.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const uuid_1 = require("uuid");
const manager_1 = tslib_1.__importDefault(require("./manager"));
class CallHierarchyManager extends manager_1.default {
    register(selector, provider) {
        return this.addProvider({
            id: (0, uuid_1.v4)(),
            selector,
            provider
        });
    }
    async prepareCallHierarchy(document, position, token) {
        let item = this.getProvider(document);
        if (!item)
            return null;
        let { provider } = item;
        return await Promise.resolve(provider.prepareCallHierarchy(document, position, token));
    }
    async provideCallHierarchyOutgoingCalls(document, item, token) {
        let providerItem = this.getProvider(document);
        if (!providerItem)
            return null;
        let { provider } = providerItem;
        return await Promise.resolve(provider.provideCallHierarchyOutgoingCalls(item, token));
    }
    async provideCallHierarchyIncomingCalls(document, item, token) {
        let providerItem = this.getProvider(document);
        if (!providerItem)
            return null;
        let { provider } = providerItem;
        return await Promise.resolve(provider.provideCallHierarchyIncomingCalls(item, token));
    }
}
exports.default = CallHierarchyManager;
