// coc-helper/tests/coc.nvim/lib/__tests__/client/workspaceFolder.test.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const assert = tslib_1.__importStar(require("assert"));
const workspaceFolders_1 = require("../../language-client/workspaceFolders");
const client_1 = require("../../language-client/client");
const vscode_uri_1 = require("vscode-uri");
class TestLanguageClient extends client_1.BaseLanguageClient {
    createMessageTransports() {
        throw new Error('Method not implemented.');
    }
    onRequest() {
        return {
            dispose: () => { }
        };
    }
}
class TestWorkspaceFoldersFeature extends workspaceFolders_1.WorkspaceFoldersFeature {
    sendInitialEvent(currentWorkspaceFolders) {
        super.sendInitialEvent(currentWorkspaceFolders);
    }
    initializeWithFolders(currentWorkspaceFolders) {
        super.initializeWithFolders(currentWorkspaceFolders);
    }
}
function testEvent(initial, then, added, removed) {
    const client = new TestLanguageClient('foo', 'bar', {});
    let arg;
    let spy = jest.spyOn(client, 'sendNotification').mockImplementation((_p1, p2) => {
        arg = p2;
        return Promise.resolve();
    });
    const feature = new TestWorkspaceFoldersFeature(client);
    feature.initializeWithFolders(initial);
    feature.sendInitialEvent(then);
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledTimes(1);
    const notification = arg;
    assert.deepEqual(notification.event.added, added);
    assert.deepEqual(notification.event.removed, removed);
}
function testNoEvent(initial, then) {
    const client = new TestLanguageClient('foo', 'bar', {});
    let spy = jest.spyOn(client, 'sendNotification').mockImplementation(() => {
        return Promise.resolve();
    });
    const feature = new TestWorkspaceFoldersFeature(client);
    feature.initializeWithFolders(initial);
    feature.sendInitialEvent(then);
    expect(spy).toHaveBeenCalledTimes(0);
}
describe('Workspace Folder Feature Tests', () => {
    const removedFolder = { uri: vscode_uri_1.URI.parse('file://xox/removed').toString(), name: 'removedName', index: 0 };
    const addedFolder = { uri: vscode_uri_1.URI.parse('file://foo/added').toString(), name: 'addedName', index: 0 };
    const addedProto = { uri: 'file://foo/added', name: 'addedName' };
    const removedProto = { uri: 'file://xox/removed', name: 'removedName' };
    test('remove/add', async () => {
        assert.ok(!client_1.MessageTransports.is({}));
        testEvent([removedFolder], [addedFolder], [addedProto], [removedProto]);
    });
    test('remove', async () => {
        testEvent([removedFolder], [], [], [removedProto]);
    });
    test('remove2', async () => {
        testEvent([removedFolder], undefined, [], [removedProto]);
    });
    test('add', async () => {
        testEvent([], [addedFolder], [addedProto], []);
    });
    test('add2', async () => {
        testEvent(undefined, [addedFolder], [addedProto], []);
    });
    test('noChange1', async () => {
        testNoEvent([addedFolder, removedFolder], [addedFolder, removedFolder]);
    });
    test('noChange2', async () => {
        testNoEvent([], []);
    });
    test('noChange3', async () => {
        testNoEvent(undefined, undefined);
    });
});
