// coc-helper/tests/coc.nvim/lib/completion/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceType = exports.CompleteFinishKind = exports.SortMethod = exports.InsertMode = void 0;
var InsertMode;
(function (InsertMode) {
    InsertMode["Insert"] = "insert";
    InsertMode["Repalce"] = "replace";
})(InsertMode = exports.InsertMode || (exports.InsertMode = {}));
var SortMethod;
(function (SortMethod) {
    SortMethod["None"] = "none";
    SortMethod["Alphabetical"] = "alphabetical";
    SortMethod["Length"] = "length";
})(SortMethod = exports.SortMethod || (exports.SortMethod = {}));
var CompleteFinishKind;
(function (CompleteFinishKind) {
    CompleteFinishKind["Normal"] = "";
    CompleteFinishKind["Confirm"] = "confirm";
    CompleteFinishKind["Cancel"] = "cancel";
})(CompleteFinishKind = exports.CompleteFinishKind || (exports.CompleteFinishKind = {}));
var SourceType;
(function (SourceType) {
    SourceType[SourceType["Native"] = 0] = "Native";
    SourceType[SourceType["Remote"] = 1] = "Remote";
    SourceType[SourceType["Service"] = 2] = "Service";
})(SourceType = exports.SourceType || (exports.SourceType = {}));
