// coc-helper/tests/coc.nvim/lib/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.watchmanCommand = exports.pluginRoot = exports.userConfigFile = exports.dataHome = exports.configHome = exports.CONFIG_FILE_NAME = exports.floatHighlightGroup = exports.APIVERSION = exports.isVim = exports.VERSION = exports.ASCII_END = void 0;
const package_json_1 = require("../../package.json");
const index_1 = require("./index");
const node_1 = require("./node");
exports.ASCII_END = 128;
exports.VERSION = package_json_1.version;
exports.isVim = process.env.VIM_NODE_RPC == '1';
exports.APIVERSION = 34;
exports.floatHighlightGroup = 'CocFloating';
exports.CONFIG_FILE_NAME = 'coc-settings.json';
exports.configHome = (0, index_1.defaultValue)(process.env.COC_VIMCONFIG, node_1.path.join(node_1.os.homedir(), '.vim'));
exports.dataHome = (0, index_1.defaultValue)(process.env.COC_DATA_HOME, node_1.path.join(node_1.os.homedir(), '.config/coc'));
exports.userConfigFile = node_1.path.join(node_1.path.normalize(exports.configHome), exports.CONFIG_FILE_NAME);
exports.pluginRoot = (0, index_1.getConditionValue)(node_1.path.dirname(__dirname), node_1.path.resolve(__dirname, '../..'));
exports.watchmanCommand = 'watchman';
