// coc-helper/tests/coc.nvim/lib/util/is.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.typedArray = exports.emptyObject = exports.objectLiteral = exports.func = exports.array = exports.number = exports.string = exports.boolean = exports.isCompletionList = exports.isCompletionItem = exports.isMarkdown = exports.isCommand = exports.isUrl = void 0;
const url_1 = require("url");
/* eslint-disable id-blacklist */
const hasOwnProperty = Object.prototype.hasOwnProperty;
function isUrl(url) {
    try {
        new url_1.URL(url);
        return true;
    }
    catch (e) {
        return false;
    }
}
exports.isUrl = isUrl;
function isCommand(obj) {
    if (!obj || !string(obj.title) || !string(obj.command) || obj.command.length == 0)
        return false;
    return true;
}
exports.isCommand = isCommand;
function isMarkdown(content) {
    if (content != null && content['kind'] == 'markdown') {
        return true;
    }
    return false;
}
exports.isMarkdown = isMarkdown;
function isCompletionItem(obj) {
    return obj && typeof obj.label === 'string';
}
exports.isCompletionItem = isCompletionItem;
function isCompletionList(obj) {
    return !Array.isArray(obj) && Array.isArray(obj.items);
}
exports.isCompletionList = isCompletionList;
function boolean(value) {
    return typeof value === 'boolean';
}
exports.boolean = boolean;
function string(value) {
    return typeof value === 'string';
}
exports.string = string;
function number(value) {
    return typeof value === 'number';
}
exports.number = number;
function array(array) {
    return Array.isArray(array);
}
exports.array = array;
function func(value) {
    return typeof value == 'function';
}
exports.func = func;
function objectLiteral(obj) {
    return (obj != null &&
        typeof obj === 'object' &&
        !Array.isArray(obj) &&
        !(obj instanceof RegExp) &&
        !(obj instanceof Date));
}
exports.objectLiteral = objectLiteral;
function emptyObject(obj) {
    if (!objectLiteral(obj)) {
        return false;
    }
    for (let key in obj) {
        if (hasOwnProperty.call(obj, key)) {
            return false;
        }
    }
    return true;
}
exports.emptyObject = emptyObject;
function typedArray(value, check) {
    return Array.isArray(value) && value.every(check);
}
exports.typedArray = typedArray;
