// coc-helper/tests/coc.nvim/lib/util/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inspect = exports.promisify = exports.unidecode = exports.fastDiff = exports.stripAnsi = exports.net = exports.vm = exports.semver = exports.which = exports.minimatch = exports.glob = exports.child_process = exports.readline = exports.debounce = exports.styles = exports.crypto = exports.os = exports.path = exports.fs = void 0;
const util_1 = require("util");
Object.defineProperty(exports, "promisify", { enumerable: true, get: function () { return util_1.promisify; } });
Object.defineProperty(exports, "inspect", { enumerable: true, get: function () { return util_1.inspect; } });
exports.fs = require('fs');
exports.path = require('path');
exports.os = require('os');
exports.crypto = require('crypto');
exports.styles = require('ansi-styles');
exports.debounce = require('debounce');
exports.readline = require('readline');
exports.child_process = require('child_process');
exports.glob = require('glob');
exports.minimatch = require('minimatch');
exports.which = require('which');
exports.semver = require('semver');
exports.vm = require('vm');
exports.net = require('net');
exports.stripAnsi = require('strip-ansi');
exports.fastDiff = require('fast-diff');
exports.unidecode = require('unidecode');
