// coc-helper/tests/coc.nvim/lib/util/platform.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWeb = exports.isNative = exports.isLinux = exports.isMacintosh = exports.isWindows = exports.platform = exports.getPlatform = exports.Platform = void 0;
var Platform;
(function (Platform) {
    Platform[Platform["Web"] = 0] = "Web";
    Platform[Platform["Mac"] = 1] = "Mac";
    Platform[Platform["Linux"] = 2] = "Linux";
    Platform[Platform["Windows"] = 3] = "Windows";
    Platform[Platform["Unknown"] = 4] = "Unknown";
})(Platform = exports.Platform || (exports.Platform = {}));
function getPlatform(process) {
    let { platform } = process;
    if (platform === 'win32')
        return Platform.Windows;
    if (platform === 'darwin')
        return Platform.Mac;
    if (platform === 'linux')
        return Platform.Linux;
    return Platform.Unknown;
}
exports.getPlatform = getPlatform;
let _platform = getPlatform(process);
exports.platform = _platform;
exports.isWindows = _platform === Platform.Windows;
exports.isMacintosh = _platform === Platform.Mac;
exports.isLinux = _platform === Platform.Linux;
exports.isNative = true;
exports.isWeb = false;
