// coc-helper/tests/coc.nvim/lib/model/db.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const node_1 = require("../util/node");
const object_1 = require("../util/object");
class DB {
    constructor(filepath) {
        this.filepath = filepath;
    }
    /**
     * Get data by key.
     *
     * @param {string} key unique key allows dot notation.
     * @returns {any}
     */
    fetch(key) {
        let obj = this.load();
        if (!key)
            return obj;
        let parts = key.split('.');
        for (let part of parts) {
            if (typeof obj[part] === 'undefined') {
                return undefined;
            }
            obj = obj[part];
        }
        return obj;
    }
    /**
     * Check if key exists
     *
     * @param {string} key unique key allows dot notation.
     */
    exists(key) {
        let obj = this.load();
        let parts = key.split('.');
        for (let part of parts) {
            if (typeof obj[part] === 'undefined') {
                return false;
            }
            obj = obj[part];
        }
        return true;
    }
    /**
     * Delete data by key
     *
     * @param {string} key unique key allows dot notation.
     */
    delete(key) {
        let obj = this.load();
        let origin = obj;
        let parts = key.split('.');
        let len = parts.length;
        for (let i = 0; i < len; i++) {
            if (typeof obj[parts[i]] === 'undefined') {
                break;
            }
            if (i == len - 1) {
                delete obj[parts[i]];
                node_1.fs.writeFileSync(this.filepath, JSON.stringify(origin, null, 2), 'utf8');
                break;
            }
            obj = obj[parts[i]];
        }
    }
    /**
     * Save data with key
     *
     * @param {string} key unique string that allows dot notation.
     * @param {number|null|boolean|string|{[index} data saved data.
     */
    push(key, data) {
        let origin = (0, object_1.toObject)(this.load());
        let obj = origin;
        let parts = key.split('.');
        let len = parts.length;
        for (let i = 0; i < len; i++) {
            let key = parts[i];
            if (i == len - 1) {
                obj[key] = data;
                let dir = node_1.path.dirname(this.filepath);
                node_1.fs.mkdirSync(dir, { recursive: true });
                node_1.fs.writeFileSync(this.filepath, JSON.stringify(origin, null, 2));
                break;
            }
            if (typeof obj[key] == 'undefined') {
                obj[key] = {};
                obj = obj[key];
            }
            else {
                obj = obj[key];
            }
        }
    }
    load() {
        let dir = node_1.path.dirname(this.filepath);
        let exists = node_1.fs.existsSync(dir);
        if (!exists) {
            node_1.fs.mkdirSync(dir, { recursive: true });
            node_1.fs.writeFileSync(this.filepath, '{}', 'utf8');
            return {};
        }
        try {
            let content = node_1.fs.readFileSync(this.filepath, 'utf8');
            return JSON.parse(content.trim());
        }
        catch (e) {
            node_1.fs.writeFileSync(this.filepath, '{}', 'utf8');
            return {};
        }
    }
    /**
     * Empty db file.
     */
    clear() {
        let exists = node_1.fs.existsSync(this.filepath);
        if (!exists)
            return;
        node_1.fs.writeFileSync(this.filepath, '{}', 'utf8');
    }
    /**
     * Remove db file.
     */
    destroy() {
        if (node_1.fs.existsSync(this.filepath)) {
            node_1.fs.unlinkSync(this.filepath);
        }
    }
}
exports.default = DB;
