// coc-helper/tests/coc.nvim/lib/model/terminal.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerminalModel = void 0;
class TerminalModel {
    constructor(cmd, args, nvim, _name, strictEnv) {
        this.cmd = cmd;
        this.args = args;
        this.nvim = nvim;
        this._name = _name;
        this.strictEnv = strictEnv;
        this.pid = 0;
    }
    async start(cwd, env) {
        let { nvim } = this;
        let cmd = [this.cmd, ...this.args];
        let [bufnr, pid] = await nvim.call('coc#terminal#start', [cmd, cwd, env || {}, !!this.strictEnv]);
        this.bufnr = bufnr;
        this.pid = pid;
    }
    onExit(code) {
        this.exitStatus = { code: code === -1 ? undefined : code };
    }
    get name() {
        return this._name || this.cmd;
    }
    get processId() {
        return Promise.resolve(this.pid);
    }
    sendText(text, addNewLine = true) {
        if (!this.bufnr)
            return;
        this.nvim.call('coc#terminal#send', [this.bufnr, text, addNewLine], true);
    }
    async show(preserveFocus) {
        let { bufnr, nvim } = this;
        if (!bufnr)
            return false;
        let [loaded, curr, winids] = await nvim.eval(`[bufloaded(${bufnr}),win_getid(),win_findbuf(${bufnr})]`);
        if (!loaded)
            return false;
        let winid = winids[0];
        if (winid && curr == winid)
            return true;
        nvim.pauseNotification();
        if (!winid) {
            nvim.command(`below ${bufnr}sb`, true);
            nvim.command('resize 8', true);
            nvim.call('coc#util#do_autocmd', ['CocTerminalOpen'], true);
        }
        else {
            nvim.call('win_gotoid', [winid], true);
        }
        nvim.command('normal! G', true);
        if (preserveFocus) {
            nvim.command('wincmd p', true);
        }
        await nvim.resumeNotification();
        return true;
    }
    async hide() {
        let { bufnr, nvim } = this;
        if (!bufnr)
            return;
        await nvim.eval(`coc#window#close(bufwinid(${bufnr}))`);
    }
    dispose() {
        if (!this.exitStatus) {
            this.exitStatus = { code: undefined };
        }
        let { bufnr, nvim } = this;
        if (!bufnr)
            return;
        this.bufnr = undefined;
        nvim.call('coc#terminal#close', [bufnr], true);
    }
}
exports.TerminalModel = TerminalModel;
