// coc-helper/tests/coc.nvim/lib/configuration/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationUpdateTarget = exports.ConfigurationTarget = void 0;
var ConfigurationTarget;
(function (ConfigurationTarget) {
    ConfigurationTarget[ConfigurationTarget["Default"] = 0] = "Default";
    ConfigurationTarget[ConfigurationTarget["User"] = 1] = "User";
    ConfigurationTarget[ConfigurationTarget["Workspace"] = 2] = "Workspace";
    ConfigurationTarget[ConfigurationTarget["WorkspaceFolder"] = 3] = "WorkspaceFolder";
    ConfigurationTarget[ConfigurationTarget["Memory"] = 4] = "Memory";
})(ConfigurationTarget = exports.ConfigurationTarget || (exports.ConfigurationTarget = {}));
var ConfigurationUpdateTarget;
(function (ConfigurationUpdateTarget) {
    ConfigurationUpdateTarget[ConfigurationUpdateTarget["Global"] = 1] = "Global";
    ConfigurationUpdateTarget[ConfigurationUpdateTarget["Workspace"] = 2] = "Workspace";
    ConfigurationUpdateTarget[ConfigurationUpdateTarget["WorkspaceFolder"] = 3] = "WorkspaceFolder";
})(ConfigurationUpdateTarget = exports.ConfigurationUpdateTarget || (exports.ConfigurationUpdateTarget = {}));
