// coc-helper/tests/coc.nvim/lib/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const vscode_uri_1 = require("vscode-uri");
const commands_1 = tslib_1.__importDefault(require("./commands"));
const manager_1 = tslib_1.__importDefault(require("./diagnostic/manager"));
const events_1 = tslib_1.__importDefault(require("./events"));
const extension_1 = tslib_1.__importDefault(require("./extension"));
const languages_1 = tslib_1.__importStar(require("./languages"));
const basic_1 = tslib_1.__importDefault(require("./list/basic"));
const manager_2 = tslib_1.__importDefault(require("./list/manager"));
const download_1 = tslib_1.__importDefault(require("./model/download"));
const fetch_1 = tslib_1.__importDefault(require("./model/fetch"));
const floatFactory_1 = tslib_1.__importDefault(require("./model/floatFactory"));
const highligher_1 = tslib_1.__importDefault(require("./model/highligher"));
const mru_1 = tslib_1.__importDefault(require("./model/mru"));
const relativePattern_1 = tslib_1.__importDefault(require("./model/relativePattern"));
const services_1 = tslib_1.__importStar(require("./services"));
const manager_3 = tslib_1.__importDefault(require("./snippets/manager"));
const string_1 = require("./snippets/string");
const sources_1 = tslib_1.__importDefault(require("./completion/sources"));
const ansiparse_1 = require("./util/ansiparse");
const errors_1 = require("./util/errors");
const mutex_1 = require("./util/mutex");
const protocol_1 = require("./util/protocol");
const window_1 = tslib_1.__importDefault(require("./window"));
const workspace_1 = tslib_1.__importDefault(require("./workspace"));
const language_client_1 = require("./language-client");
const line_1 = tslib_1.__importDefault(require("./model/line"));
const semanticTokensBuilder_1 = require("./model/semanticTokensBuilder");
const index_1 = require("./tree/index");
const util_1 = require("./util");
const fs_1 = require("./util/fs");
const processes_1 = require("./util/processes");
const types_1 = require("./configuration/types");
const types_2 = require("./completion/types");
const workspaceFolder_1 = require("./core/workspaceFolder");
module.exports = {
    get nvim() {
        return workspace_1.default.nvim;
    },
    Uri: vscode_uri_1.URI,
    LineBuilder: line_1.default,
    NullLogger: language_client_1.NullLogger,
    SettingMonitor: language_client_1.SettingMonitor,
    LanguageClient: language_client_1.LanguageClient,
    CancellationTokenSource: protocol_1.CancellationTokenSource,
    ProgressType: protocol_1.ProgressType,
    RequestType: protocol_1.RequestType,
    RequestType0: protocol_1.RequestType0,
    NotificationType: protocol_1.NotificationType,
    NotificationType0: protocol_1.NotificationType0,
    ProtocolRequestType: protocol_1.ProtocolRequestType,
    ProtocolRequestType0: protocol_1.ProtocolRequestType0,
    ProtocolNotificationType: protocol_1.ProtocolNotificationType,
    ProtocolNotificationType0: protocol_1.ProtocolNotificationType0,
    Highligher: highligher_1.default,
    Mru: mru_1.default,
    Emitter: protocol_1.Emitter,
    SnippetString: string_1.SnippetString,
    BasicList: basic_1.default,
    Mutex: mutex_1.Mutex,
    TreeItem: index_1.TreeItem,
    SemanticTokensBuilder: semanticTokensBuilder_1.SemanticTokensBuilder,
    FloatFactory: floatFactory_1.default,
    RelativePattern: relativePattern_1.default,
    CancellationError: errors_1.CancellationError,
    WorkspaceChange: vscode_languageserver_types_1.WorkspaceChange,
    ResponseError: protocol_1.ResponseError,
    Trace: protocol_1.Trace,
    DocumentUri: vscode_languageserver_types_1.DocumentUri,
    WorkspaceFolder: vscode_languageserver_types_1.WorkspaceFolder,
    InlineValueText: vscode_languageserver_types_1.InlineValueText,
    InlineValueVariableLookup: vscode_languageserver_types_1.InlineValueVariableLookup,
    InlineValueEvaluatableExpression: vscode_languageserver_types_1.InlineValueEvaluatableExpression,
    InlineValueContext: vscode_languageserver_types_1.InlineValueContext,
    InlayHintKind: vscode_languageserver_types_1.InlayHintKind,
    InlayHintLabelPart: vscode_languageserver_types_1.InlayHintLabelPart,
    InlayHint: vscode_languageserver_types_1.InlayHint,
    DiagnosticRelatedInformation: vscode_languageserver_types_1.DiagnosticRelatedInformation,
    SemanticTokens: vscode_languageserver_types_1.SemanticTokens,
    SemanticTokenTypes: vscode_languageserver_types_1.SemanticTokenTypes,
    SemanticTokenModifiers: vscode_languageserver_types_1.SemanticTokenModifiers,
    AnnotatedTextEdit: vscode_languageserver_types_1.AnnotatedTextEdit,
    ChangeAnnotation: vscode_languageserver_types_1.ChangeAnnotation,
    SymbolTag: vscode_languageserver_types_1.SymbolTag,
    Command: vscode_languageserver_types_1.Command,
    Color: vscode_languageserver_types_1.Color,
    CodeDescription: vscode_languageserver_types_1.CodeDescription,
    ColorInformation: vscode_languageserver_types_1.ColorInformation,
    ColorPresentation: vscode_languageserver_types_1.ColorPresentation,
    TextDocumentEdit: vscode_languageserver_types_1.TextDocumentEdit,
    TextDocumentIdentifier: vscode_languageserver_types_1.TextDocumentIdentifier,
    VersionedTextDocumentIdentifier: vscode_languageserver_types_1.VersionedTextDocumentIdentifier,
    TextDocumentItem: vscode_languageserver_types_1.TextDocumentItem,
    DocumentHighlight: vscode_languageserver_types_1.DocumentHighlight,
    SelectionRange: vscode_languageserver_types_1.SelectionRange,
    DocumentLink: vscode_languageserver_types_1.DocumentLink,
    CodeLens: vscode_languageserver_types_1.CodeLens,
    FormattingOptions: vscode_languageserver_types_1.FormattingOptions,
    CodeAction: vscode_languageserver_types_1.CodeAction,
    CodeActionContext: vscode_languageserver_types_1.CodeActionContext,
    DocumentSymbol: vscode_languageserver_types_1.DocumentSymbol,
    WorkspaceSymbol: vscode_languageserver_types_1.WorkspaceSymbol,
    CreateFile: vscode_languageserver_types_1.CreateFile,
    RenameFile: vscode_languageserver_types_1.RenameFile,
    WorkspaceEdit: vscode_languageserver_types_1.WorkspaceEdit,
    InsertReplaceEdit: vscode_languageserver_types_1.InsertReplaceEdit,
    InsertTextMode: vscode_languageserver_types_1.InsertTextMode,
    CompletionItem: vscode_languageserver_types_1.CompletionItem,
    CompletionList: vscode_languageserver_types_1.CompletionList,
    Hover: vscode_languageserver_types_1.Hover,
    ParameterInformation: vscode_languageserver_types_1.ParameterInformation,
    SignatureInformation: vscode_languageserver_types_1.SignatureInformation,
    SymbolInformation: vscode_languageserver_types_1.SymbolInformation,
    MarkupContent: vscode_languageserver_types_1.MarkupContent,
    ErrorCodes: protocol_1.ErrorCodes,
    CompletionItemTag: vscode_languageserver_types_1.CompletionItemTag,
    integer: vscode_languageserver_types_1.integer,
    uinteger: vscode_languageserver_types_1.uinteger,
    FoldingRangeKind: vscode_languageserver_types_1.FoldingRangeKind,
    FoldingRange: vscode_languageserver_types_1.FoldingRange,
    ChangeAnnotationIdentifier: vscode_languageserver_types_1.ChangeAnnotationIdentifier,
    DeleteFile: vscode_languageserver_types_1.DeleteFile,
    OptionalVersionedTextDocumentIdentifier: vscode_languageserver_types_1.OptionalVersionedTextDocumentIdentifier,
    CompletionItemLabelDetails: vscode_languageserver_types_1.CompletionItemLabelDetails,
    MarkedString: vscode_languageserver_types_1.MarkedString,
    ProviderName: languages_1.ProviderName,
    DocumentDiagnosticReportKind: protocol_1.DocumentDiagnosticReportKind,
    UniquenessLevel: protocol_1.UniquenessLevel,
    MonikerKind: protocol_1.MonikerKind,
    PatternType: workspaceFolder_1.PatternType,
    SourceType: types_2.SourceType,
    ConfigurationTarget: types_1.ConfigurationUpdateTarget,
    ServiceStat: services_1.ServiceStat,
    FileType: fs_1.FileType,
    State: language_client_1.State,
    ClientState: language_client_1.ClientState,
    CloseAction: language_client_1.CloseAction,
    ErrorAction: language_client_1.ErrorAction,
    TransportKind: language_client_1.TransportKind,
    MessageTransports: language_client_1.MessageTransports,
    RevealOutputChannelOn: language_client_1.RevealOutputChannelOn,
    MarkupKind: vscode_languageserver_types_1.MarkupKind,
    DiagnosticTag: vscode_languageserver_types_1.DiagnosticTag,
    DocumentHighlightKind: vscode_languageserver_types_1.DocumentHighlightKind,
    SymbolKind: vscode_languageserver_types_1.SymbolKind,
    SignatureHelpTriggerKind: protocol_1.SignatureHelpTriggerKind,
    FileChangeType: protocol_1.FileChangeType,
    CodeActionKind: vscode_languageserver_types_1.CodeActionKind,
    Diagnostic: vscode_languageserver_types_1.Diagnostic,
    DiagnosticSeverity: vscode_languageserver_types_1.DiagnosticSeverity,
    CompletionItemKind: vscode_languageserver_types_1.CompletionItemKind,
    InsertTextFormat: vscode_languageserver_types_1.InsertTextFormat,
    Location: vscode_languageserver_types_1.Location,
    LocationLink: vscode_languageserver_types_1.LocationLink,
    CancellationToken: protocol_1.CancellationToken,
    Position: vscode_languageserver_types_1.Position,
    Range: vscode_languageserver_types_1.Range,
    TextEdit: vscode_languageserver_types_1.TextEdit,
    Disposable: protocol_1.Disposable,
    Event: protocol_1.Event,
    workspace: workspace_1.default,
    window: window_1.default,
    CodeActionTriggerKind: vscode_languageserver_types_1.CodeActionTriggerKind,
    CompletionTriggerKind: protocol_1.CompletionTriggerKind,
    snippetManager: manager_3.default,
    events: events_1.default,
    services: services_1.default,
    commands: commands_1.default,
    sources: sources_1.default,
    languages: languages_1.default,
    diagnosticManager: manager_1.default,
    extensions: extension_1.default,
    listManager: manager_2.default,
    TreeItemCollapsibleState: index_1.TreeItemCollapsibleState,
    terminate: processes_1.terminate,
    fetch: fetch_1.default,
    download: download_1.default,
    ansiparse: ansiparse_1.ansiparse,
    disposeAll: util_1.disposeAll,
    concurrent: util_1.concurrent,
    watchFile: fs_1.watchFile,
    wait: util_1.wait,
    runCommand: processes_1.runCommand,
    isRunning: processes_1.isRunning,
    executable: processes_1.executable,
};
